### ISSUE TITLE
RedisCommandException for "Command cannot be issued to a replica" is difficult to handle

### ISSUE THREAD

**Issue by AdamOutcalt** [StackExchange/StackExchange.Redis#2016]
Hey all,

I work on a large service using Azure Redis Premium. Our normal pattern for using Redis is something like:

try{
  db.StringGet()
}
catch (RedisException)
...
catch (TimeoutException)
...

This works pretty well, but we have some low level of failures that look like this:

RedisCommandException: Exception: 'Command cannot be issued to a slave: UNLINK <key>'

(In newer versions of the SDK, it says "Command cannot be issued to a replica" instead.)

This is a reasonable error, but I am not intentionally targeting a replica. I assume Azure has failed over from under me, changing the recipient to a replica, while it was primary when the command was sent.

I would like to just treat this as a transient failure, log it, and move on with life as though it were a timeout. This is a little tricky though because the exception is RedisCommandException (which doesn't inherit from RedisException). I am hesitant to catch and handle RedisCommandException, because I understand that it usually means the command is bad, like maybe a bad LUA script.

Basically, I'd like some way to differentiate this exception so I can treat it as transient.

My suggestion is to make this case a RedisException or RedisServerException, but I'm open to other resolutions, including being told that I should be catching some different set of exceptions as transient.

Thanks,
Adam

**Comment by NickCraver** [comment]
Which version of the library are you using currently?

**Comment by AdamOutcalt** [comment]
2.2.79

**Comment by AdamOutcalt** [comment]
2.2.79 (currently shows "replica" in the text)

**Comment by hao9638422** [comment]
Same issue here:
StackExchange.Redis.RedisCommandException: Command cannot be issued to a replica: SETEX xxx

**Comment by pseidel-kcf** [comment]
Having this issue with UNLINK on AWS clustered Elasticache 6.x using StackExchange 2.6.45 (2.6.48 in our staging environment does not appear to resolve the issue). Explicitly adding `DemandMaster` as a parameter to `KeyDeleteAsync` mitigates the issue.

**Comment by slorello89** [comment]
@NickCraver 

Looks like the pattern of checking features and passing in the server might be proving to be a bit problematic.

Looks like the issue is propagating from [here](https://github.com/StackExchange/StackExchange.Redis/blob/main/src/StackExchange.Redis/RedisDatabase.cs#L759)

When the feature check is done [here](https://github.com/StackExchange/StackExchange.Redis/blob/main/src/StackExchange.Redis/RedisDatabase.cs#L776), there's nothing that allows the writeability of the command to influence server-selection, so it's just a crap-shoot whether you get back a master or replica to pass further down the line.

Need to add a command parameter to `GetFeatures`

**Comment by adomass** [comment]
We are still getting StackExchange.Redis.RedisCommandException: Command cannot be issued to a replica: HMSET
with latest version 2.6.66. The fix for #2191 did not help... We use Redis not only as cache but also as [hangfire](https://github.com/marcoCasamento/Hangfire.Redis.StackExchange) storage and this issue causes problems for scheduling jobs.
